packages/core/src/linker/view_container_ref.ts

```diff
@@ -148,11 +168,266 @@ export abstract class ViewContainerRef {
+      if (prevIdx !== -1) {
+        this.detach(prevIdx);
+      } else {
+        const prevLContainer = lView[PARENT] as LContainer;
+        ngDevMode &&
+            assertEqual(
+                isLContainer(prevLContainer), true,
+                'An attached view should have its PARENT point to a container.');
+
+
+        // We need to re-create a R3ViewContainerRef instance since those are not stored on
+        // LView (nor anywhere else).
+        const prevVCRef = new R3ViewContainerRef(
+            prevLContainer, prevLContainer[T_HOST] as TDirectiveHostNode, prevLContainer[PARENT]);
+
+        prevVCRef.detach(prevVCRef.indexOf(viewRef));
+      }
+    }
+
+    // Logical operation of adding `LView` to `LContainer`
+    const adjustedIdx = this._adjustIndex(index);
+    const lContainer = this._lContainer;
+    insertView(tView, lView, lContainer, adjustedIdx);
+
+    // Physical operation of adding the DOM nodes.
+    const beforeNode = getBeforeNodeForView(adjustedIdx, lContainer);
+    const renderer = lView[RENDERER];
+    const parentRNode = nativeParentNode(renderer, lContainer[NATIVE] as RElement | RComment);
+    if (parentRNode !== null) {
+      addViewToContainer(tView, lContainer[T_HOST], renderer, lView, parentRNode, beforeNode);
+    }
+
+    (viewRef as R3ViewRef<any>).attachToViewContainerRef(this);
+    addToArray(lContainer[VIEW_REFS]!, adjustedIdx, viewRef);
+
+    return viewRef;
+  }
+
+  move(viewRef: ViewRef, newIndex: number): ViewRef {
+    if (viewRef.destroyed) {
+      throw new Error('Cannot move a destroyed View in a ViewContainer!');
+    }
+    return this.insert(viewRef, newIndex);
+  }
+
+  indexOf(viewRef: ViewRef): number {
+    const viewRefsArr = this._lContainer[VIEW_REFS];
+    return viewRefsArr !== null ? viewRefsArr.indexOf(viewRef) : -1;
+  }
+
+  remove(index?: number): void {
+    this.allocateContainerIfNeeded();
+    const adjustedIdx = this._adjustIndex(index, -1);
+    const detachedView = detachView(this._lContainer, adjustedIdx);
+
+    if (detachedView) {
+      // Before destroying the view, remove it from the container's array of `ViewRef`s.
+      // This ensures the view container length is updated before calling
+      // `destroyLView`, which could recursively call view container methods that
+      // rely on an accurate container length.
+      // (e.g. a method on this view container being called by a child directive's OnDestroy
+      // lifecycle hook)
+      removeFromArray(this._lContainer[VIEW_REFS]!, adjustedIdx);
+      destroyLView(detachedView[TVIEW], detachedView);
+    }
+  }
+
+  detach(index?: number): ViewRef|null {
+    this.allocateContainerIfNeeded();
+    const adjustedIdx = this._adjustIndex(index, -1);
+    const view = detachView(this._lContainer, adjustedIdx);
+
+    const wasDetached = view && removeFromArray(this._lContainer[VIEW_REFS]!, adjustedIdx) != null;
+    return wasDetached ? new R3ViewRef(view!) : null;
+  }
```

Comment:
I'd like to propose a small refactoring here:
```suggestion
  detach(index?: number): ViewRef|null {
    const lContainer = this._lContainer;
    const adjustedIdx = this._adjustIndex(index, -1);
    const view = detachView(lContainer, adjustedIdx);

    const viewRefs = getOrCreateViewRefs(lContainer);  // <-- previously this.allocateContainerIfNeeded()
    const wasDetached = view && removeFromArray(viewRefs, adjustedIdx) != null;
    return wasDetached ? new R3ViewRef(view!) : null;
  }
```

and the `getOrCreateViewRefs` (replacement to `allocateContainerIfNeeded`) might looks like:

```
  function getOrCreateViewRefs(lContainer: LContainer): ViewRef[] {
    if (lContainer[VIEW_REFS] === null) {
      lContainer[VIEW_REFS] = [];
    }
    return lContainer[VIEW_REFS];
  }
```

This is outside of the main scope of this PR, so we can do that later too.